src/atomicdex/api/mm2/mm2.cpp

```diff
@@ -34,6 +42,121 @@ namespace
             answer.result = j.at("result").get<RpcSuccessReturnType>();
         }
     }
+
+    std::pair<QString, QString>
+    extract_error(const nlohmann::json& events, const QStringList& error_events)
+    {
+        for (auto&& cur_event: events)
+        {
+            if (std::any_of(std::begin(error_events), std::end(error_events), [&cur_event](auto&& error_str) {
+                    return cur_event.at("state").get<std::string>() == error_str.toStdString();
+                }))
+            {
+                //! It's an error
+                if (cur_event.contains("data") && cur_event.at("data").contains("error"))
+                {
+                    return {
+                        QString::fromStdString(cur_event.at("state").get<std::string>()),
+                        QString::fromStdString(cur_event.at("data").at("error").get<std::string>())};
+                }
+            }
+        }
+        return {};
+    }
+
+    QString
+    determine_order_status_from_last_event(const nlohmann::json& events, const QStringList& error_events) noexcept
+    {
+        if (events.empty())
+        {
+            return "matching";
+        }
+        auto last_event = events.back().at("state").get<std::string>();
+        if (last_event == "Started")
+        {
+            return "matched";
+        }
+
+        if (last_event == "TakerPaymentWaitRefundStarted" || last_event == "MakerPaymentWaitRefundStarted")
+        {
+            return "refunding";
+        }
+
+        QString status = "ongoing";
+
+        if (last_event == "Finished")
+        {
+            status = "successful";
+            //! Find error or not
+            for (auto&& cur_event: events)
+            {
+                if (cur_event.contains("data") && cur_event.at("data").contains("error") &&
+                    std::any_of(std::begin(error_events), std::end(error_events), [&cur_event](auto&& error_str) {
+                        return cur_event.at("state").get<std::string>() == error_str.toStdString();
+                    }))
+                {
+                    status = "failed";
+                }
+            }
+        }
+
+        return status;
```

Comment:
In general this function works very well but having constant values ​​directly in the code is disturbing, the API could change and it would be more difficult to find the potential problems. Using constant values ​​in the API file would make it easier to maintain errors. Not mandatory for this split, but this is suboptimal.